src/coreclr/jit/emit.h

```diff
@@ -2173,22 +2173,71 @@ class emitter
         return (emitCurIG && emitCurIGfreeNext > emitCurIGfreeBase);
     }
 
-    instrDesc* emitLastIns;
+#define EMIT_MAX_IG_INS_COUNT 256
+
+    // A recording of the last instructions, used for peephole optimizations.
+    // The max count of last instructions is the same as the max count of IG instructions.
+    instrDesc* emitLastInstrs[EMIT_MAX_IG_INS_COUNT];
+
+    inline instrDesc* emitGetLastIns() const
+    {
+        return emitLastInstrs[emitInsCount % ArrLen(emitLastInstrs)];
+    }
+
+    inline bool emitHasLastIns() const
+    {
+        return (emitGetLastIns() != nullptr);
+    }
 
     // Check if a peephole optimization involving emitLastIns is safe.
     //
     // We must have a lastInstr to consult.
     // The emitForceNewIG check here prevents peepholes from crossing nogc boundaries.
     // The final check prevents looking across an IG boundary unless we're in an extension IG.
-    bool emitCanPeepholeLastIns()
+    bool emitCanPeepholeLastIns() const
     {
-        return (emitLastIns != nullptr) &&                 // there is an emitLastInstr
-               !emitForceNewIG &&                          // and we're not about to start a new IG
+        return emitHasLastIns() && !emitForceNewIG &&      // and we're not about to start a new IG
                ((emitCurIGinsCnt > 0) ||                   // and we're not at the start of a new IG
                 ((emitCurIG->igFlags & IGF_EXTEND) != 0)); //    or we are at the start of a new IG,
                                                            //    and it's an extension IG
     }
 
+    enum emitPeepholeResult
+    {
+        PEEPHOLE_CONTINUE,
+        PEEPHOLE_ABORT
+    };
+
+    // Visits the last emitted instructions.
+    // Must be safe to do - use emitCanPeepholeLastIns for checking.
+    template <typename Action>
+    void emitPeepholeLastInstrs(Action action)
+    {
+        assert(emitCanPeepholeLastIns());
+
+        instrDesc* id;
+
+        // We use 'emitCurIGinsCnt' because we do not want to reference an instruction
+        // from a previous IG as to protect from GC holes.
+        for (unsigned i = 0; i < emitCurIGinsCnt; i++)
```

Comment:
Because this is based on `emitCurIGinsCnt`, it doesn't allow peeking back from IGF_EXTEND groups to previous groups.

We should allow that, since overflow groups are somewhat arbitrarily sized.

One way to do that might be to create a new `emitCurIGinsCntExtended` (or some other name) variable that counts all the instructions in a non-IGF_EXTEND IG and all subsequent IGF_EXTEND IGs (thus, is zeroed out when we create a new non-IGF_EXTEND IG).

Reply:
We should do this in a follow-up PR to avoid making this one more complicated; but I agree we should definitely do this.

Reply:
Since I added `emitCurLastInsCnt` , it behaves like you said for `emitCurIGinsCntExtended`. However, actual asm diffs are small when the `emitLastInstrs` includes instructions from the current IG when creating a new extended IG. But, at least there won't be inconsistencies.